File: extensions/widgets/interactive/ImageClickInput/ImageClickInput.py
# coding: utf-8
#
# Copyright 2014 The Oppia Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, softwar
# distributed under the License is distributed on an "AS-IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

__author__ = 'Zhan Xiong Chin'

from core.domain import widget_domain
from extensions.value_generators.models import generators

class ImageClickInput(widget_domain.BaseWidget):
    """Interactive widget for multiple choice selection on an image"""

    # The human-readable name of the widget.
    name = 'Image'

    # The category the widget falls under in the widget repository.
    category = 'Custom'

    # A description of the widget.
    description = 'A widget where users click on regions of an image.'

    # Customization parameters and their descriptions, schemas and default
    # values.
    _customization_arg_specs = [{
        'name': 'image_and_regions',
        'description': 'The image and its regions.',
        'schema': {
            'type': 'custom',
            'obj_type': 'SegmentedImage',
        },
        'default_value': {
            'imagePath': '',
            'imageRegions': []
        },
    }]

    # Actions that the reader can perform on this widget which trigger a
    # feedback interaction, and the associated input types. Interactive widgets
    # must have at least one of these. This attirbute name MUST be prefixed by
    # '_'.
    _handlers = [{
        'name': 'submit', 'obj_type': 'CoordTwoDim'   
    }]

    # Additional JS library dependencies that should be loaded in pages
    # containing this widget. These should correspond to names of files in
    # feconf.DEPENDENCIES_TEMPLATES_DIR.
    _dependency_ids = []

File: extensions/widgets/interactive/ImageClickInput/ImageClickInput.js
// Copyright 2014 The Oppia Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS-IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Directive for the ImageClickInput interactive widget.
 *
 * IMPORTANT NOTE: The naming convention for customization args that are passed
 * into the directive is: the name of the parameter, followed by 'With',
 * followed by the name of the arg.
 */

oppia.directive('oppiaInteractiveImageClickInput', [
  '$rootScope', '$sce', 'oppiaHtmlEscaper', function($rootScope, $sce, oppiaHtmlEscaper) {
    return {
      restrict: 'E',
      scope: {},
      templateUrl: 'interactiveWidget/ImageClickInput',
      controller: ['$scope', '$element', '$attrs', function($scope, $element, $attrs) {
        var imageAndRegions = oppiaHtmlEscaper.escapedJsonToObj($attrs.imageAndRegionsWithValue);
        $scope.filepath = imageAndRegions.imagePath;
        $scope.imageUrl = $sce.trustAsResourceUrl(
          '/imagehandler/' + $rootScope.explorationId + '/' +
          encodeURIComponent($scope.filepath)
        );
        $scope.onClickImage = function(event) {
          var image = $($element).find('.oppia-image-click-img');
          var mouseX = (event.pageX - image.offset().left) / image.width();
          var mouseY = (event.pageY - image.offset().top) / image.height();
          var answer = imageAndRegions.imageRegions.length;
          for (var i = 0; i < imageAndRegions.imageRegions.length; i++) {
            var region = imageAndRegions.imageRegions[i].region;
            if (region[0][0] <= mouseX && mouseX <= region[1][0] &&
                region[0][1] <= mouseY && mouseY <= region[1][1]) {
              answer = i;
              break;
            }
          }
          $scope.$parent.$parent.submitAnswer(answer, 'submit');
        }
      }]
    };
  }
]);
